sway_libs/src/fixed_point/ufp/ufp64/ufp64.sw

```diff
@@ -0,0 +1,267 @@
+        Self { value }
+    }
+
+    /// The smallest value that can be represented by this type.
+    pub fn min() -> Self {
+        Self {
+            value: ~u64::min(),
+        }
+    }
+
+    /// The largest value that can be represented by this type,
+    pub fn max() -> Self {
+        Self {
+            value: ~u64::max(),
+        }
+    }
+
+    /// The size of this type in bits.
+    pub fn bits() -> u32 {
+        64
+    }
+}
+
+impl core::ops::Eq for UFP64 {
+    fn eq(self, other: Self) -> bool {
+        self.value == other.value
+    }
+}
+
+impl core::ops::Ord for UFP64 {
+    fn gt(self, other: Self) -> bool {
+        self.value > other.value
+    }
+
+    fn lt(self, other: Self) -> bool {
+        self.value < other.value
+    }
+}
+
+impl core::ops::Add for UFP64 {
+    /// Add a UFP64 to a UFP64. Panics on overflow.
+    fn add(self, other: Self) -> Self {
+        Self {
+            value: self.value + other.value,
+        }
+    }
+}
+
+impl core::ops::Subtract for UFP64 {
+    /// Subtract a UFP64 from a UFP64. Panics of overflow.
+    fn subtract(self, other: Self) -> Self {
+        // If trying to subtract a larger number, panic.
+        assert(self.value >= other.value);
+
+        Self {
+            value: self.value - other.value,
+        }
+    }
+}
+
+impl core::ops::Multiply for UFP64 {
+    /// Multiply a UFP64 with a UFP64. Panics of overflow.
+    fn multiply(self, other: Self) -> Self {
+        let self_u128 = ~U128::from(0, self.value);
+        let other_u128 = ~U128::from(0, other.value);
+
+        let self_multiply_other = self_u128 * other_u128;
+        let res_u128 = self_multiply_other >> 32;
+        if res_u128.upper != 0 {
+            // panic on overflow
+            revert(0);
+        }
+
+        Self {
+            value: res_u128.lower,
+        }
+    }
+}
+
+impl core::ops::Divide for UFP64 {
+    /// Divide a UFP64 by a UFP64. Panics if divisor is zero.
+    fn divide(self, divisor: Self) -> Self {
+        let zero = ~UFP64::zero();
+        assert(divisor != zero);
+
+        let denominator = ~U128::from(0, ~Self::denominator());
+        // Conversion to U128 done to ensure no overflow happen
+        // and maximal precision is avaliable
+        // as it makes possible to multiply by the denominator in 
+        // all cases
+        let self_u128 = ~U128::from(0, self.value);
+        let divisor_u128 = ~U128::from(0, divisor.value);
+
+        // Multiply by denominator to ensure accuracy 
+        let res_u128 = self_u128 * denominator / divisor_u128;
+
+        if res_u128.upper != 0 {
+            // panic on overflow
+            revert(0);
+        }
+        Self {
+            value: res_u128.lower,
+        }
+    }
+}
+
+impl UFP64 {
+    /// Creates UFP64 that correponds to a unsigned integer
+    pub fn from_uint(uint: u64) -> Self {
+        Self {
+            value: ~Self::denominator() * uint,
+        }
+    }
+}
+
+impl UFP64 {
```

Comment:
There are multiple instances of `impl UFP64`. Could these be condensed into one for readability?

Reply:
Unfortunately no, due to this known and AFAIK yet not fixed issue